### PR TITLE
Implement matrix3

### PR DESCRIPTION
**Pull request by XiaokaiUber**
- implement APIs in matrix3 class
- add test cases
- add doc

### CODE REVIEW

**Comment by ibgreen** [comment]
@tsherif FYI

File: src/matrix3.js [hunk 1]
```diff
@@ -19,24 +19,21 @@
...
 
 import MathArray from './lib/math-array';
-// import {checkNumber} from './lib/common';
+import {checkNumber} from './lib/common';
 import Vector2 from './vector2';
 import Vector3 from './vector3';
-// import Vector2, {validateVector2} from './vector2';
-// import Vector3, {validateVector3} from './vector3';
-// import assert from 'assert';
+import {validateVector} from './matrix4';
```
Comment (ibgreen) [inline src/matrix3.js]: We should minimize imports between similar classes. Either copy this function or move it to a utility file.

File: src/matrix3.js [hunk 2]
```diff
@@ -191,86 +173,56 @@ export default class Matrix3 extends MathArray {
...
         return this.transformVector3(vector, out);
-      // case 4:
-      //   return this.transformVector4(vector, out);
       default:
         throw new Error('Illegal vector');
     }
   }
 
   transformDirection(vector, out) {
+    return this._transformVector(vector, out, 0);
```
Comment (ibgreen) [inline src/matrix3.js]: the transformDirection/transformPoint distinction will only apply if this matrix was used with 2D coordinates. We should either cut these functions (recommended for now) or implement them correctly.

File: src/matrix3.js [hunk 3]
```diff
@@ -191,86 +173,56 @@ export default class Matrix3 extends MathArray {
...
-        vec4.transformMat4(tempVector4, [vector[0], vector[1], vector[2], 0], this);
+        if (Boolean(w) !== Boolean(vector[2])) {
+          throw new Error('math.gl: Matrix3.transformPoint - invalid vector');
+        }
         out = out || new Vector3();
-        [out[0], out[1], out[2]] = tempVector4;
+        vec3.transformMat3(out, vector, this);
+        validateVector(out, 3);
         break;
-      // case 4:
```
Comment (ibgreen) [inline src/matrix3.js]: Would could easily support vector4's. We would just treat this matrix as if it was a 4x4 matrix with this matrix as the upper left "minor": 

```
| mat3 0 |
| 0    1 |
```

File: src/matrix4.js [hunk 4]
```diff
@@ -54,7 +54,7 @@ export function validateMatrix4(m) {
   );
 }
 
-function validateVector(v, length) {
+export function validateVector(v, length) {
```
Comment (ibgreen) [inline src/matrix4.js]: See comment above. Let's minimize cross-class imports.

File: src/matrix3.js (hunk 1 above)
Comment (XiaokaiUber) [reply src/matrix3.js]: `validateVector` is moved to a new file `src/lib/validators.js`

**Comment by XiaokaiUber** [comment]
> Very nice!!!
> 
> It would be cool to see some bundle size measurements for the new class. You can test this using `yarn build ; yarn metrics`. You may need to add your new class to `test/size` directory.

Here is the info of `matrix3` by running `yarn metrics`:
Bundle Size: 13KB (13793) KB 
Compressed: 3KB (3961) KB

File: src/matrix3.js [hunk 5]
```diff
@@ -73,40 +61,54 @@ export default class Matrix3 extends MathArray {
   }
 
   /* eslint-disable max-params */
-  setRowMajor(m00 = 1, m10 = 0, m20 = 0, m01 = 0, m11 = 1, m21 = 0, m02 = 0, m12 = 0, m22 = 1) {
-    return this.set(m00, m01, m02, m10, m11, m12, m20, m21, m22);
+  setRowMajor(m00 = 1, m01 = 0, m02 = 0, m10 = 0, m11 = 1, m12 = 0, m20 = 0, m21 = 0, m22 = 1) {
```
Comment (Pessimistress) [inline src/matrix3.js]: Names of the arguments should match their position in the matrix.

File: test/src/matrix3.spec.js [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+
+  tapeEquals(t, m, RESULT, 'copy gave the right result');
+
+  t.end();
+});
+
+test('Matrix3#fromQuaternion', t => {
+  t.equals(typeof Matrix3.prototype.fromQuaternion, 'function');
+
+  const RESULT = [
```
Comment (Pessimistress) [inline test/src/matrix3.spec.js]: Where is the expected result from? Add a comment on how future maintainers can validate its correctness.

**Comment by ibgreen** [comment]
> Here is the info of `matrix3` by running `yarn metrics`:

Thanks. This is about half the size of matrix4 which is what I expected. Always good to keep an eye on these numbers, if we add inappropriate dependencies they can quickly shoot up.

```
| 2.3.0          | es6  | 13KB (13793) KB  | 3KB (3961) KB     | matrix3.js
| 2.3.0          | es6  | 21KB (22103) KB  | 6KB (6489) KB     | matrix4.js
```

File: src/matrix3.js (hunk 5 above)
Comment (XiaokaiUber) [reply src/matrix3.js]: I just compared it with the implementation in `matrix4`. It seems the naming in the `setColumnMajor` is wrong. In `setRowMajor`, the matrix passed in should follow the row major indices. in `setColumnMajor`, it follows the column major indices.

**Pull request merged**